frontend/catalyst/debug/compiler_functions.py

```diff
@@ -160,3 +160,39 @@ def compile_from_mlir(ir, compiler=None, compile_options=None):
         result_types = [mlir.ir.RankedTensorType.parse(rt) for rt in func_data[1].split(",")]
 
     return CompiledFunction(shared_object, qfunc_name, result_types, None, compiler.options)
+
+
+@debug_logger
+def get_pipeline_output(fn, pass_name):
+    """Capture IR string from the given compiler pass.
+
+    Args:
+        fn (QJIT): a qjit-decorated function
+        pass_name (str): target compiler pass name
+
+    Returns:
+        str: output ir from the target compiler pass
+    """
+    if not fn.compiler.options.keep_intermediate:
+        raise RuntimeError("keep_intermediate must be set to True to get pipeline's output")
+    if pass_name == "last":
+        return fn.compiler.last_compiler_output.get_output_ir()
+    return fn.compiler.get_output_of(pass_name)
+
+
+@debug_logger
+def replace_ir(fn, pass_name, new_ir):
+    """Specify new IR that will be used for future compilation.
+
+    Args:
+        fn (QJIT): a qjit-decorated function
+        pass_name (str): name of a pass. Recompilation happens after his pass.
+        new_ir (str): new ir in the string format
+    """
+    if pass_name and new_ir:
+        fn.overwrite_ir = new_ir
```

Comment:
Do we want to overwrite the current QJIT, or do we want to clone a debugging QJIT and set the ir of the clone? I have no preference towards wither option but I would love to know the rationale behind overwriting.

Reply:
``fn.overwrite_ir`` is to allow us to compile from MLIR/LLVM or we have to go through Jax.  Since we want to modify MLIR/LLVM directly, we have to find a way to skip jax. (I think the original compilation flow is ``python function`` -> ``jax`` -> ``mlir/llvm``.)

I though cloning QJIT is trivial so I only handle overwriting here.